### ISSUE TITLE
Custom DI wont work as it forces use of IServiceProvider

### ISSUE THREAD

**Issue by MatthewSwarbrick** [danielgerlag/workflow-core#116]
I have a web api project in dotnet core 2.0 and I'm using simple injector as my IoC container. 
In startup.cs I am using the simple injector's crosswire feature which hands over responsibility of resolving services to simple injector.
I cross wired IWorkflowHost so I could resolve it in my controllers, this works fine, but I had a custom step defined in my workflow. As I haven't registered this step in the serviceCollection it fails to construct the step body. I would expect as I'm using simple injector for this to use simple injector to resolve the dependency, however because it uses the IServiceProvider to get the registered services it can't find it. 
Have I missed something? I think currently I would have to register all my dependencies twice to get this working, one for simple injector, and one for the service provider. I don't want to be doing that

**Comment by danielgerlag** [comment]
`IServiceProvider` is simply an abstraction for an IoC container.  Most .NET IoC libraries have added support for this abstraction in the form of a bridge package of some kind.  I'm not sure about `SimpleInjector` though.  Have you thought of replacing `SimpleInjector` with something else?

**Comment by cnayan** [comment]
Maybe useful:

I decided to let a `Step` be created with `IServiceProvider` in ctor. So, I can use the ServiceProvider and work with my registered objects.

https://github.com/cnayan/workflow-core/blob/7b1fead4825de01c1d2a6d333c7ca9c9e9963f77/src/WorkflowCore/Models/WorkflowStep.cs#L51

**Comment by cnayan** [comment]
Correct! However, would `ConstructBody` not be restrictive still, in the ctor call? (check the link to the code.)

**Comment by danielgerlag** [comment]
I have no idea what you are trying to do with this

**Comment by cnayan** [comment]
:-) No worries.

**Comment by Miggleness** [comment]
I think what @cnayan  is trying to achieve is eliminating the need to register steps.

I do understand the overhead as in my case, I create multiple projects for steps which can add to maintenance overhead if I need to add each of them.  His approach takes this maintenance out of the picture in favor of a service locator.

I've taken the approach below to minimize the need to register steps (if you have multiple steps per project).

```
private static void AddSteps(this IServiceCollection services, Assembly[] assemblies)
        {
            var addedAssembly = new HashSet<Assembly>();

            foreach(var a in assemblies)
            {
                if (addedAssembly.Contains(a))
                    continue;

                addedAssembly.Add(a);

                // add step body
                foreach (var type in a.GetTypes()
                    .Where(t=> (typeof(StepBodyAsync).IsAssignableFrom(t) || typeof(StepBody).IsAssignableFrom(t)))
                    .Where(t=> t.IsClass)
                    .Where(t=> !t.IsAbstract))
                {
                    services.AddScoped(type);
                }
            }
        }
```
